### PR TITLE
ENH Add a conda-based CI job on azure.

### LINKED ISSUE #18711: Add a conda-based CI job
Suggested in https://github.com/numpy/numpy/pull/18659#issuecomment-804248214 after adding an `environment.yml` file. However, this is useful more broadly - making sure things build with conda (e.g. could catch `numpy.disutils` issues) is a more compelling reason to do this than exercising a yaml file.

### PR DESCRIPTION
**Pull request by cmarmo**
This pull request adds a conda based job on azure, using the `environment.yml` file.
Thanks for considering it.

Closes #18711.

Please note that I can reproduce the failing test locally, in a pip virtual environment.
But I can't see the same failure in the other azure builds: I'm afraid this is beyond my understanding... :confused: 
 ```py
path = '/home/cmarmo/software/numpy/build/testenv/lib64/python3.9/site-packages/numpy/typing/tests/data/reveal/testing.py'
reveal = 'tuple[types.ModuleType]'
expected_reveal = "32: note: Revealed type is 'builtins.tuple[_importlib_modulespec.ModuleType]'", lineno = 32

    def _test_reveal(path: str, reveal: str, expected_reveal: str, lineno: int) -> None:
        if reveal not in expected_reveal:
>           raise AssertionError(_REVEAL_MSG.format(lineno, expected_reveal, reveal))
E           AssertionError: Reveal mismatch at line 32
E           
E           Expected reveal: "32: note: Revealed type is 'builtins.tuple[_importlib_modulespec.ModuleType]'"
E           Observed reveal: 'tuple[types.ModuleType]'

expected_reveal = "32: note: Revealed type is 'builtins.tuple[_importlib_modulespec.ModuleType]'"
lineno     = 32
path       = '/home/cmarmo/software/numpy/build/testenv/lib64/python3.9/site-packages/numpy/typing/tests/data/reveal/testing.py'
reveal     = 'tuple[types.ModuleType]'

numpy/typing/tests/test_typing.py:295: AssertionError
=============================================== short test summary info ===============================================
FAILED numpy/typing/tests/test_typing.py::test_reveal[testing.py] - AssertionError: Reveal mismatch at line 32
```

### CODE REVIEW

**Comment by BvB93** [comment]
The issue is that mypy is still pinned to an old version:
https://github.com/numpy/numpy/blob/764662504a233ce218543aae3257dafdaa39e4f0/environment.yml#L21

General question: Is there a way for dependabot to update the  `environment.yml` in addition to `test_requirements.txt`?
If not, it might be better to just remove the pin.

**Comment by cmarmo** [comment]
> The issue is that mypy is still pinned to an old version:
> 
> https://github.com/numpy/numpy/blob/764662504a233ce218543aae3257dafdaa39e4f0/environment.yml#L21

Thanks! Indeed, I had `mypy == 0.812` in my environment....

**Comment by mattip** [comment]
Perhaps a different approach, since it seems dependabot does not yet know about conda, would be to parse the requirements.txt and the environment.yml and make sure they match.

**Review by rgommers — COMMENTED** [review]
This is great, thanks for working on this @cmarmo!

The contents of the job look good. I just have one suggestion: given that the output of the conda install is >2500 lines, could you separate the `steps:` into two `script:` stages with different display names? Then it'll be easier to see if something failed during install or during test suite run, and the log will be easier to read through.

Similar to https://github.com/numpy/numpy/blob/a3fb80213bc28b6ac1e6a7da09a2af3149a677aa/azure-pipelines.yml#L55

> General question: Is there a way for dependabot to update the environment.yml in addition to test_requirements.txt?
If not, it might be better to just remove the pin.

There's not, AFAIK. Removing the pin sounds fine to me.

**Comment by rgommers** [comment]
> Perhaps a different approach, since it seems dependabot does not yet know about conda, would be to parse the requirements.txt and the environment.yml and make sure they match.

Please let's not do anything that complicated.

**Comment by mattip** [comment]
Linking to the dependabot issue for visibility dependabot/dependabot-core#2227. Although that is only part of the story. Only a specific test for equivalence would give us some level of confidence that the spec is the same. @rgommers why do you think that is more complicated than maintaining yet another CI run?

**Comment by rgommers** [comment]
> Only a specific test for equivalence would give us some level of confidence that the spec is the same. @rgommers why do you think that is more complicated than maintaining yet another CI run?

That's not quite the right question. Adding a CI job is very useful, because it detects failing tests and build/packaging issues that otherwise may show up after making a release or on a new contributor's local test suite run.

Adding some kind of parsing to compare `requirements.txt` doesn't help that much on the other hand. When we update `mypy`, either via dependabot or some other way, tests usually fail and the real work is fixing them. Having to then bump the version by hand also in `environment.yml` is very simple, and this CI job will already remind you if you forget (as the one current failure here shows). So given that there's little to gain here, it doesn't seem useful to spend extra effort to get this right and maintain it.

Another point: `environment.yml` and `test_requirements.txt` shouldn't match exactly, since they don't have the exact same purpose (you cannot set up a dev env based on just `test_requirements.txt`). 

tl;dr let's do nothing here unless it turns out there's a real need.

> Removing the pin sounds fine to me.

Sorry, now that I've typed the above, let me change this: the pin should be updated to `mypy==0.902`, not removed. Because that will ensure CI fails once we intentionally upgrade mypy and not whenever there's a new mypy release. Other requirements are unpinned and that's usually fine, but Mypy has a habit of breaking things all the time.

**Comment by cmarmo** [comment]
Thanks for your answers and clarifications.
I have split the pipeline in two pieces.
Also I realized that azure ubuntu already come with conda, there is no need to download install miniconda then.
Finally, to keep the log shorter and to be consistent with the other pipelines, I have removed `set -x`.

**Review by rgommers — APPROVED** [review]
Looks great, thanks @cmarmo.

This is very slightly wasteful, but that's fine (it's a "set up a dev env" `environment.yml` rather than a "only what we need for CI" one):
```
qt-5.12.9            | 99.5 MB
scipy-1.7.0          | 22.0 MB
```

> Also I realized that azure ubuntu already comes with conda

Very nice - didn't know that. Thank you Microsoft:)
